### PR TITLE
Check wallet balance in CreateServerConn

### PR DESCRIPTION
**Pull request by billfort**
1. Add minBalance in entry and exit config;
2. Check wallet balance in tuna.CreateServerConn. If wallet balance is less than config minBalance, connection would be rejected.

### CODE REVIEW

File: config.go [hunk 1]
```diff
@@ -130,6 +133,7 @@ type ExitConfiguration struct {
 	TcpDialContext                 func(ctx context.Context, network, addr string) (net.Conn, error) `json:"-"`
 	HttpDialContext                func(ctx context.Context, network, addr string) (net.Conn, error) `json:"-"`
 	WsDialContext                  func(ctx context.Context, network, addr string) (net.Conn, error) `json:"-"`
+	MinBalance                     float64                                                           `json:"minBalance"`
```
Comment (yilunzhang) [inline config.go:136]: Should be `ReverseMinBalance` because it's used in reverse mode

File: tuna.go [hunk 2]
```diff
@@ -745,6 +750,21 @@ func (c *Common) CreateServerConn(force bool) error {
 				return err
 			}
 
+			entryToExitMaxPrice, exitToEntryMaxPrice, err := ParsePrice(c.ServiceInfo.MaxPrice)
+			if err != nil {
+				return err
+			}
+			if entryToExitMaxPrice > 0 || exitToEntryMaxPrice > 0 {
+				balance, err := c.Wallet.Balance()
```
Comment (yilunzhang) [inline tuna.go:758]: Better to use `c.Client.BalanceByAddress(c.Wallet.Address())` because it's more reliable

File: config.go [hunk 3]
```diff
@@ -72,6 +73,7 @@ type EntryConfiguration struct {
 	TcpDialContext                   func(ctx context.Context, network, addr string) (net.Conn, error) `json:"-"`
 	HttpDialContext                  func(ctx context.Context, network, addr string) (net.Conn, error) `json:"-"`
 	WsDialContext                    func(ctx context.Context, network, addr string) (net.Conn, error) `json:"-"`
+	MinBalance                       float64                                                           `json:"minBalance"`
```
Comment (yilunzhang) [inline config.go:76]: Better to use string type

File: tuna.go [hunk 4]
```diff
@@ -146,6 +146,8 @@ type Common struct {
 
 	reverseBytesExitToEntry map[string][]uint64
 	reverseBytesEntryToExit map[string][]uint64
+
+	minBalance float64 // minimum wallet balance requirement for connecting
```
Comment (yilunzhang) [inline tuna.go:150]: Use `Fixed64` type and parse it with `StringToFixed64` method

File: tuna.go [hunk 5]
```diff
@@ -745,6 +750,21 @@ func (c *Common) CreateServerConn(force bool) error {
...
+			if err != nil {
+				return err
+			}
+			if entryToExitMaxPrice > 0 || exitToEntryMaxPrice > 0 {
+				balance, err := c.Wallet.Balance()
+				if err != nil {
+					return err
+				}
+				fBalance, _ := strconv.ParseFloat(balance.Fixed64.String(), 64)
+				if fBalance <= c.minBalance {
```
Comment (yilunzhang) [inline tuna.go:763]: Should be `<`

File: tuna.go [hunk 6]
```diff
@@ -745,6 +750,21 @@ func (c *Common) CreateServerConn(force bool) error {
 				return err
 			}
 
+			entryToExitMaxPrice, exitToEntryMaxPrice, err := ParsePrice(c.ServiceInfo.MaxPrice)
```
Comment (yilunzhang) [inline tuna.go:753]: Only need to check these if `minBalance > 0`

File: tuna.go [hunk 7]
```diff
@@ -745,6 +750,21 @@ func (c *Common) CreateServerConn(force bool) error {
...
+			if err != nil {
+				return err
+			}
+			if entryToExitMaxPrice > 0 || exitToEntryMaxPrice > 0 {
+				balance, err := c.Wallet.Balance()
+				if err != nil {
+					return err
```
Comment (yilunzhang) [inline tuna.go:760]: I think we shouldn't return when failed to get balance so tuna won't be affected by on-chain incidents.

File: config.go (hunk 1 above)
Comment (billfort) [reply config.go:136]: Yes, modified

File: tuna.go (hunk 2 above)
Comment (billfort) [reply tuna.go:758]: Yes, modified.

File: config.go (hunk 3 above)
Comment (billfort) [reply config.go:76]: Yes, modified.

File: tuna.go (hunk 4 above)
Comment (billfort) [reply tuna.go:150]: Yes, modified.

File: tuna.go (hunk 5 above)
Comment (billfort) [reply tuna.go:763]: Yes, modified.

File: tuna.go (hunk 6 above)
Comment (billfort) [reply tuna.go:753]: Yes, modified.

File: tuna.go (hunk 7 above)
Comment (billfort) [reply tuna.go:760]: Yes, modified.

File: tuna.go [hunk 8]
```diff
@@ -250,6 +253,7 @@ func NewCommon(
 		udpWriteChan:  make(chan []byte),
 		connReadyChan: make(map[string]chan struct{}),
 	}
+	c.minBalance, _ = common.StringToFixed64(minBalance)
```
Comment (yilunzhang) [inline tuna.go:256]: Always handle error

File: tuna.go [hunk 9]
```diff
@@ -745,6 +749,19 @@ func (c *Common) CreateServerConn(force bool) error {
...
 			}
 
+			if c.minBalance > 0 {
+				entryToExitMaxPrice, exitToEntryMaxPrice, err := ParsePrice(c.ServiceInfo.MaxPrice)
+				if err != nil {
+					return err
+				}
+				if entryToExitMaxPrice > 0 || exitToEntryMaxPrice > 0 {
+					balance, err := c.Client.BalanceByAddress(c.Wallet.Address())
+					if err == nil && balance.Fixed64 < c.minBalance {
```
Comment (yilunzhang) [inline tuna.go:759]: Better use `ToFixed64()`

File: tuna.go (hunk 8 above)
Comment (billfort) [reply tuna.go:256]: Yes, modified.

File: tuna.go (hunk 9 above)
Comment (billfort) [reply tuna.go:759]: Yes, modified.

File: tuna.go [hunk 10]
```diff
@@ -745,6 +752,19 @@ func (c *Common) CreateServerConn(force bool) error {
...
 			}
 
+			if c.minBalance > 0 {
+				entryToExitMaxPrice, exitToEntryMaxPrice, err := ParsePrice(c.ServiceInfo.MaxPrice)
+				if err != nil {
+					return err
+				}
+				if entryToExitMaxPrice > 0 || exitToEntryMaxPrice > 0 {
+					balance, err := c.Client.BalanceByAddress(c.Wallet.Address())
+					if err == nil && balance.ToFixed64() < c.minBalance {
```
Comment (yilunzhang) [inline tuna.go:762]: I would suggest we log the error if not nil

Comment (billfort) [reply tuna.go:762]: Yes, modified.

**Review by yilunzhang — APPROVED** [review]

**Pull request merged**